---
title: "<b>Omix</b> - Integrative analysis of Multi-omics data in R"
author: "This vignette was created by ElÃ©onore Schneegans and Nurun Fancy"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 3
    code_folding: show
    df_print: paged
vignette: >
  %\VignetteIndexEntry{<b>Omix</b> - Integrative analysis of Multi-omics data in R}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
editor_options: 
  markdown: 
  wrap: 72
---
```{r style, echo=FALSE, results='asis', message=FALSE}
BiocStyle::markdown()
knitr::opts_chunk$set(tidy = FALSE,
                        message = FALSE)
options(knitr.duplicate.label = "allow")
```

```{r setup}
library(Omix)
```

## Overview

The goal of Omix is to provide tools in R to build a complete analysis workflow 
for integrative analysis of data generated from multi-omics platform.

-   Generate a multi-omics object using `MultiAssayExperiment`.

-   Quality control of single-omics data.

-   Formatting, normalisation, denoising of single-omics data.

-   Separate single-omics analyses.

-   Integration of multi-omics data for combined analysis. 

-   Publication quality plots and interactive analysis reports based of shinyApp.

## Running *Omix*

The `Omix` pipeline requires the following input:

-   rawdata_rna: a data-frame containing raw RNA count with rownames as gene and colnames as samples.

-   rawdata_protein:  a data-frame containing protein abundance with rownames as proteins and colnames as samples.

-   map_rna: A data-frame of two columns named `primary` and `colname` where primary should contain unique sample name with a link to sample metadata and colname is the column names of the `rawdata_rna` data-frame.

-   map_protein: A data-frame of two columns named `primary` and `colname` where primary should contain unique sample name with a link to sample metadata and colname is the column names of the `rawdata_protein` data-frame. 

-   metadata_rna: A data-frame containing `rna` assay specific metadata where rownames are same as the colnames of `rawdata_rna` data.frame.

-   metadata_protein: A data-frame containing `protein` assay specific metadata where rownames are same as the colnames of `rawdata_protein` data.frame.

-   sample_metadata: A data-frame containing sample level metadata for both omics assays.


### Step one - Generate MultiAssayExperiment object

Call required packages for vignette:
```{r}
library(Omix)
```

```{r}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install('proBatch','mixOmix')
```


First we must get the data from Omix for the vignette:

```{r}
rawdata_rna <- readRDS('~/RDS_ukdri/multiomics/Omix_test_env/data/rawdata_rna.rds')
rawdata_protein <- readRDS('~/RDS_ukdri/multiomics/Omix_test_env/data/rawdata_protein.rds')
map_rna <- readRDS('~/RDS_ukdri/multiomics/Omix_test_env/data/map_rna.rds')
map_protein <- readRDS('~/RDS_ukdri/multiomics/Omix_test_env/data/map_protein.rds')
metadata_rna<- readRDS('~/RDS_ukdri/multiomics/Omix_test_env/data/metadata_rna.rds')
metadata_protein <- readRDS('~/RDS_ukdri/multiomics/Omix_test_env/data/metadata_protein.rds')
individual_metadata <- readRDS('~/RDS_ukdri/multiomics/Omix_test_env/data/individual_metadata.rds')
rna_qc_data_matrix <- readRDS('~/RDS_ukdri/multiomics/Omix_test_env/data/rna_qc_data_matrix.rds')
```

Input files structure

# Raw rna counts

Raw RNA is a data frame of raw counts, with features as rows and samples as columns

```{r}
print(rawdata_rna)
```

# Raw protein abundance

Raw protein is a data frame of raw counts, with features as rows and samples as columns

```{r}
print(rawdata_protein)
```
# Maps

Maps are data frame containing two columns: `primary` and `colname`. 
The `primary` column should be the individual ID the individual metadata, and the 
`colname` the matched sample names from the raw matrices columns. If sample and
individual ids are the same, maps aren't needed (primary and colnames are the same).
```{r}
print(map_protein)
print(map_rna)
```
# Technical metadata

Technical metadata are data frames contain the column `colname` that should match
the sample names in the raw matrices, and any additional columns related to technical artefacts like batches 
```{r}
print(metadata_rna)
```
# Sample metadata

Sample metadata should contain individual level metadata, where one column matches the `primary` column in maps.
```{r}
print(sample_metadata)
```
# Optional inputs

Optional inputs contain additional data frame used for QC visualisation purposes 
```{r}
print(rna_qc_data_matrix)
```


# Generate a multiassay experiment object 

To run Omix, we first need to generate a multi-omics object

```{r}
multiassay=generate_multiassay(rawdata_rna = rawdata_rna,
                  rawdata_protein = rawdata_protein,
                  individual_to_sample=FALSE,
                  map_rna = map_rna,
                  map_protein = map_protein,
                  metadata_rna = metadata_rna,
                  metadata_protein = metadata_protein,
                  individual_metadata = individual_metadata,
                  map_by_column = 'individualID',
                  rna_qc_data=TRUE,
                  rna_qc_data_matrix=rna_qc_data_matrix,
                  organism='human')
          
```

The MultiAssayExperiment object was succesfully created. 
The following steps will process and perform QC on each omic layers of the 
`multiassay` object.

### Step two - Process transcriptomics data

```{r}
multiassay=process_rna(multiassay=multiassay,
            transformation='vst',
            protein_coding=TRUE,
            min_count = 10,
            min_sample = 0.6,
            dependent= 'diagnosis',
            levels=c('Control','Mid','Late'),
            covariates=c('pmi','msex','age_death'),
            filter=TRUE,
            batch_correction=TRUE,
            batch="libraryBatch",
            remove_sample_outliers= TRUE)
```



### Optional - Add custom processed transcriptomics slot

```{r}
multiassay=processed_rna(multiassay=multiassay,
                         custom_processed_df=custom_rna)
```

```{r}
multiassay=multiassay[,,c(TRUE,TRUE,TRUE,FALSE)]
```


### Step three - Process proteomics data


```{r}
multiassay=process_protein(
    multiassay=multiassay,
    min_sample = 0.5,
    dependent =  "diagnosis",
    levels = c('Control','Mid','Late'),
    imputation = 'minimum_value',
    remove_feature_outliers= FALSE,
    batch_correction= TRUE,
    batch="Batch",
    correction_method="Median_centering",
    remove_sample_outliers=TRUE,
    denoise=TRUE,
    covariates=c('pmi','msex','age_death'))
```

The processing parameter are stored in the object


```{r}
multiassay@metadata$parameters_processing_protein
```


### Optional - Add custom processed transcriptomics slot

```{r}
multiassay=processed_protein(multiassay=multiassay,
                         custom_processed_df=custom_protein)
```

# Processed multiassay object

```{r}
print(multiassay)
```

# Venn Diagram of raw and processed slots 

```{r}
upsetSamples(multiassay)
```

### Step Four - Generate QC dashboards 

Finally we produce a report with report_qc (this takes a few minutes): -

```{r}
#Store results in same temp dir
report_qc(multiassay, report_file = "qc_report_omix",
              report_folder_path = paste0(outputDir,"/tmp_Zeisel2015_scflow/"))
```


### Step Five - Single omic analyses 

## Transcriptomics 

Now we run a differential expression analysis between groups of interest, where 
the first level, here control is the reference. The function requires to run `process_rna`
first, with covariates of interest. 

```{r}
multiassay=rna_DE_analysis(multiassay,
                    dependent='diagnosis',
                    levels=c('Control','Mid','Late'),
                    filter_protein_coding = TRUE,
                    log2FoldChange = 0.25)
```

Differential expression results for transcripts are saved in the `multiassay@metadata$DEG` slot 

```{r}

DT::datatable(multiassay@metadata$DEG$LatevsControl,
              rownames = FALSE, 
              escape = FALSE,
              options = list(pageLength = 20, 
                             scrollX=T, 
                             autoWidth = TRUE,
                             columnDefs = list(list(width = '500px', targets = 1)),
                             dom = 'Blfrtip'))

```


Volcano plot of differentially expressed genes is saved in the `multiassay@metadata$DEG$plot` slot

```{r}
volcano_interactive(multiassay@metadata[["DEG"]]$LatevsControl)
```

```{r}
multiassay@metadata$DEG$plot
```

Pathway analysis using ORA can be performed on up and down regulated genes 

```{r}
up_genes= multiassay@metadata$DEG$sig_gene$LatevsControl$up
up_genes
pathway_analysis_enrichr_up_genes=pathway_analysis_enrichr(interest_gene = up_genes ,
                                     project_name = NULL,
                                     enrichment_database = c(
                                       "GO_Molecular_Function_2021",
                                       "GO_Cellular_Component_2021",
                                       "GO_Biological_Process_2021",
                                       "WikiPathways_2021_Human",
                                       "Reactome_2016",
                                       "KEGG_2021_Human",
                                       "MSigDB_Hallmark_2020",
                                       "BioCarta_2016"
                                     ),
                                     is_output = FALSE,
                                     output_dir = ".",
                                     plot_n = 20) 
```


```{r}
DT::datatable(pathway_analysis_enrichr_up_genes$`GO_Biological_Process_2021`,
              rownames = FALSE, 
              escape = FALSE,
              options = list(pageLength = 20, 
                             scrollX=T, 
                             autoWidth = TRUE,
                             columnDefs = list(list(width = '500px', targets = 1)),
                             dom = 'Blfrtip'))
```
The enrichment results can be visualised in a dot plot.

```{r}
pathway_analysis_enrichr_up_genes$plot$GO_Biological_Process_2021
```

## Proteomics


Now we run a differential expression analysis between groups of interest, where 
the first level, here control is the reference. If the data is already denoised for covariates of interest `process_rna`, set covariates as `NA``

```{r}
multiassay=protein_DE_analysis(multiassay,
                         slot="protein_processed",
                         dependent='diagnosis',
                         covariates=NA,
                         levels=c('Control','Mid','Late'),
                         log2FoldChange = 0.25)
```

Differential expression results for proteins are saved in the `multiassay@metadata$DEP` slot 

```{r}

DT::datatable(multiassay@metadata$DEP$LatevsControl,
              rownames = FALSE, 
              escape = FALSE,
              options = list(pageLength = 20, 
                             scrollX=T, 
                             autoWidth = TRUE,
                             columnDefs = list(list(width = '500px', targets = 1)),
                             dom = 'Blfrtip'))

```

Volcano plot of differentially abundant proteins is saved in the `multiassay@metadata$DEP$plot` slot

```{r}

volcano_interactive(multiassay@metadata[["DEP"]]$LatevsControl)
```


Pathway analysis using ORA can be performed on up and down regulated proteins

```{r}
up_proteins= multiassay@metadata$DEP$sig_protein$LatevsControl$up
pathway_analysis_enrichr_up_proteins=pathway_analysis_enrichr(interest_gene = up_proteins,
                                     project_name = NULL,
                                     enrichment_database = c(
                                       "GO_Molecular_Function_2021",
                                       "GO_Cellular_Component_2021",
                                       "GO_Biological_Process_2021",
                                       "WikiPathways_2021_Human",
                                       "Reactome_2016",
                                       "KEGG_2021_Human",
                                       "MSigDB_Hallmark_2020",
                                       "BioCarta_2016"
                                     ),
                                     is_output = FALSE,
                                     output_dir = ".",
                                     min_overlap = 3,
                                     plot_n = 10) 
```

```{r}
DT::datatable(pathway_analysis_enrichr_up_proteins$`GO_Biological_Process_2021`,
              rownames = FALSE, 
              escape = FALSE,
              options = list(pageLength = 20, 
                             scrollX=T, 
                             autoWidth = TRUE,
                             columnDefs = list(list(width = '500px', targets = 1)),
                             dom = 'Blfrtip'))
```
The enrichment results can be visualised in a dot plot.

```{r}
pathway_analysis_enrichr_up_proteins$plot$GO_Biological_Process_2021
```

```{r}
background=.get_background(multiassay, of='protein')
enrichGO=clusterProfiler::enrichGO(gene = up_proteins,
                                             OrgDb = org.Hs.eg.db,
                                             universe = background,
                                             keyType       = 'SYMBOL',
                                             ont           = "BP",
                                             pAdjustMethod = "BH",
                                             pvalueCutoff  = 0.05,
                                             qvalueCutoff  = 0.05,
                                             minGSSize = 3,
                                             maxGSSize = 500)
```


```{r}

enrichGO@result<-enrichGO@result[enrichGO@result$Count>=3,]
enrichplot::dotplot(enrichGO, showCategory=10) 
enrichplot::heatplot(enrichGO, showCategory=10)
```
```{r}
xx <- enrichplot::pairwise_termsim(enrichGO)                     
enrichplot::cnetplot(xx, circular = TRUE, colorEdge = TRUE)
```
```{r}
enrichplot::treeplot(xx)
```

## Comparing differentially expressed genes and protein 

```{r}
comparison_plot=single_omic_comparison(multiassay,slot='LatevsControl')
comparison_plot
```

### Step Five - Multi-omics integration using DIABLO

By default, the vertical integration is performed on processed slots.

DIABLO is a supervised multi-omics integration model that requires the input of a dependent variable.
The function will perform tuning on the defined range, and number of components chosen.


```{r}
multiassay=vertical_integration(multiassay,
                                integration='DIABLO',
                                intersect_genes=FALSE,
                                ID_type='gene_name',
                                dependent='Group',
                                levels= c("Control", "Case"),
                                design='cor',
                                ncomp=2,
                                range=list(mRNA=seq(50,500,by=50),
                                              proteins=seq(50,500,by=50)))
```


```{r}
multiassay=vertical_integration(multiassay,
                                integration='DIABLO',
                                intersect_genes=FALSE,
                                ID_type='gene_name',
                                dependent='diagnosis',
                                levels= c("Control", "Mid","Late"),
                                design='cor',
                                ncomp=2,
                                range=list(mRNA=seq(50,500,by=50),
                                              proteins=seq(50,500,by=50)))
```

### Step Five - Multi-omics integration using sMBPLS

```{r}

multiassay=vertical_integration(multiassay,
                                ID_type = "gene_name",
                                integration='sMBPLS',
                                dependent='gpath',
                                design='cor',
                                ncomp=2,
                                list.keepX=list(mRNA = c(50), proteins = c(50)))
```

### Step Five - Multi-omics integration using MOFA

By default, the vertical integration is performed on processed slots.
All default parameters are models default parameters. 

```{r}
multiassay=vertical_integration(multiassay,
                                integration='MOFA',
                                dependent='diagnosis',
                                num_factors = 5,
                                scale_views = TRUE)
```

 
### Step Five - Multi-omics integration using clustering (iCluster)

We will perform clustering on an a priori homogenous subgroup of patients, and try to identify "subtypes".
Here, we are interesed in early stage patients (Braak II - III)

```{r}
Early_multiassay <- multiassay[, multiassay$diagnosis == "Mid"]
```


```{r}

multiassay=vertical_integration(Early_multiassay,
                                ID_type = "gene_name",
                                integration='iCluster',
                                try.N.clust= 2:4)
```


### Step Six - Multi-omics analyses (Supervised Example using DIABLO)

The user may want to perform the integrative downstream analyses step-by-step, 
or to run `integrative_results_supervised`, which will create a results object 
containing, multi-omics signatures, network visualisation, cell type and functional 
enrichment, as well as comparison of signatures against the OpenTargets knowledge base.

```{r}
int_results=integrative_results_supervised(multiassay,
                                integration = "DIABLO 2",
                                component = 1,
                                correlation_threshold=0.5,
                                enrichment_method='enrichr',
                                disease_id='MONDO_0004975',
                                simplify=TRUE)
```

### Exploring results I - Multiomics signatures 

```{r}
int_results$detrimental$multiomics_signature
```


### Exploring results II - Multiomics networks 

Purple shows genes
Green shows protein (gene name)
Networks are generated in two direction, detrimental and protective, assuming that the first level of the provided dependent variable is the Control/ non disease group.

```{r}
.interactive_network(int_results$detrimental$network)
```

```{r}
.interactive_network(int_results$protective$network)

```

### Exploring results II b - Multiomics networks communities 

```{r}
int_results$detrimental$network_communities
```

### Exploring results II c - Multiomics networks communities to cell type 

```{r}
int_results$detrimental$cell_type_communities
```

### Exploring results III - Functional enrichment of multiomics signatures 

```{r}
int_results$detrimental$enrichment$protein$plot$GO_Biological_Process_2021
```

### Exploring results IV - Relating signatures to current biological knowledge

```{r}
int_results$detrimental$open_targets
```
### Step Seven - Multi-omics analyses (Clustering example using iCluster)

```{r}
int_results_clustering=integrative_results_clustering(multiassay=Early_multiassay,
                                           intersect_genes = FALSE,
                                           slots = c(
                                             "rna_processed",
                                             "protein_processed"
                                           ),
                                           ID_type = "gene_name",
                                           integration = "iCluster",
                                           enrichment_method='enrichr',
                                           annotations_continuous = c("amyloid","nft", "gpath"),
                                           annotations_cat = c("braaksc","progression","cogdx"),
                                           res.path = "~/RDS_ukdri/multiomics/Omix_test_env/figures_OMIX_clustering/",
                                           correlation_threshold = 0.5) 
```


```{r}
int_results_clustering$clinical
```


```{r}
int_results_clustering$DE$rna$plot
```
```{r}
int_results_clustering$DE$proteins$plot
```
```{r}

.interactive_network(int_results_clustering$Up$network$`CS1-CS2`)
```

```{r}
dotplot(int_results_clustering$comparison$`CS1-CS2`$rna)
treeplot(xx2,showCategory=15,nWords=1,fontsize=3,offset_tiplab=4,
         offset = 1.4)
         

```


```{r}
  xx <- pairwise_termsim(int_results_clustering$comparison$`CS1-CS2`$rna)                     
  emapplot(xx)
```
```{r}
simplified=clusterProfiler::simplify(int_results_clustering$comparison$`CS1-CS2`$rna, cutoff=0.6, by="p.adjust", select_fun=min)

xx2 <- pairwise_termsim(simplified)                     
treeplot(xx2,showCategory = 15)
emapplot(xx2,showCategory = 15)

```

```{r}
dotplot(simplified,showCategory = 40)
```


### Step eight - Comparing single omic & multiomic results 


- Significant features comparison
- Geneset clusters 
