<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>A Multi-Omics Integration Pipeline • Omix</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="A Multi-Omics Integration Pipeline">
<meta property="og:description" content="*Omix* provides a generalisable framework to efficiently process, integrate and analyse multi-omics data in an end-to-end manner. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">Omix</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Omix_vignette_CV_MTG.html">Omix - Get started</a>
    </li>
    <li>
      <a href="articles/Omix_vignette_MAP_CV_pseudotemporal.html">Omix - Pseudotemporal multi-omics integration</a>
    </li>
    <li>
      <a href="articles/demo.html">Omix - Sample reports</a>
    </li>
  </ul>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">

<p><img src="reference/figures/Omix_logo.png" width="200" style="align:left;"></p>
<!-- badges: start -->

<div class="section level2">
<h2 id="pipeline-overview">Pipeline overview<a class="anchor" aria-label="anchor" href="#pipeline-overview"></a>
</h2>
<p>The <code>Omix</code> pipeline offers an integration and analysis framework for multi-omics intended to pre-process, analyse, and v isualise multimodal data flexibly to address research questions. Omix is built on four consecutive blocks, (1) preparation of the multimodal container, (2) processing and quality control, (3) single omic analyses, and (4) multi-omics vertical integration.</p>
<p>Pipeline outputs are standardised and include publication-quality plots, tables, and interactive reports.</p>

<p><img src="reference/figures/omix_pipeline_n.png"  width="auto"></p>

<p>Omix offers a range of state-of-the-art processing functions and integrative models, equipped with versatile parameters and quality control features. This empowers researchers to explore different integration strategies easily, enhancing the speed, scalability, and flexibility of multi-omics analyses.</p>
<p>While the current version focuses on bulk transcriptomics and proteomics, future iterations aim to encompass a broader range of omics types, expanding the software’s applicability and usefulness.</p>
</div>
<div class="section level2">
<h2 id="installation">Installation<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>You can install the development version of Omix from <a href="https://github.com/" class="external-link">GitHub</a> with:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># install.packages("devtools")</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"eleonore-schneeg/Omix"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="overview-of-omix-features">Overview of Omix features<a class="anchor" aria-label="anchor" href="#overview-of-omix-features"></a>
</h2>
<ul>
<li>Multi-omics data container
<ul>
<li>The Omix multimodal container harmonises data management of multiple omics datasets. It enables the storage of raw and processed omics data slots, along with patient metadata, technical metadata, analysis parameters and outputs. The object structure relies on the <a href="https://bioconductor.org/packages/release/bioc/html/MultiAssayExperiment.html" class="external-link">MultiAssayExperiment library</a>
</li>
</ul>
</li>
<li>Data processing &amp; Quality Control
<ul>
<li>Each omics layer is processed separately according to best practices. Given the wide range of processing functionalities, users decide which parameters and steps of the modular sequence are performed, which involves all or a combination of the folllowing steps:</li>
<li>Feature filtering</li>
<li>Normalisation/ transformation</li>
<li>Batch correction &amp; denoising</li>
<li>Sample outlier removal</li>
<li>Formatting</li>
</ul>
</li>
<li>Single platform models
<ul>
<li>Omix provides a suite of analysis options including differential analysis (DE), a standard method to identify genes that are differentially expressed between certain disease states; <a href="https://cran.r-project.org/web/packages/WGCNA/index.html" class="external-link">Weighted Gene correlation Network (WGCNA)</a>, to identify modules of genes that associate to certain disease covariates; <a href="http://mixomics.org" class="external-link">sparse Partial Least Square (sPLS)</a> to define a sparse set of omics features, or molecular signature, that explains the response variable.</li>
</ul>
</li>
<li>Vertical integration for joint analysis
<ul>
<li>Omix provides a series of state-of-the-art integration models to perform patient-specific multi-omic integration, including:</li>
<li><a href="https://biofam.github.io/MOFA2/" class="external-link">MOFA</a></li>
<li><a href="https://biofam.github.io/MOFA2/" class="external-link">MEIFESTO</a></li>
<li><a href="http://mixomics.org" class="external-link">Sparse Multi-Block PLS (sMB-PLS)</a></li>
<li><a href="http://mixomics.org" class="external-link">DIABLO</a></li>
<li>Multi-omics clustering models such as <a href="https://rdrr.io/bioc/iClusterPlus/man/iCluster.html" class="external-link">iCluster</a>
</li>
</ul>
</li>
<li>Downstream analyses
<ul>
<li>Multi-omics integration is followed by a series of downstream analyses, including:</li>
<li>Multi-omics networks with <a href="https://igraph.org" class="external-link">iGraph</a>
</li>
<li>Community detection with the <a href="https://igraph.org" class="external-link">Louvain or Leiden clustering algorithms</a>
</li>
<li>Pseudotime inference with <a href="https://bioconductor.org/packages/release/bioc/html/slingshot.html" class="external-link">Slingshot</a>
</li>
<li>Functional enrichment with <a href="https://maayanlab.cloud/Enrichr/" class="external-link">EnrichR</a>
</li>
<li>Cell-type enrichment with <a href="https://bioconductor.org/packages/release/bioc/html/EWCE.html" class="external-link">EWCE</a>
</li>
<li>Target validation based on the <a href="https://www.opentargets.org" class="external-link">OpenTargets</a> database</li>
<li>Publication quality plots and analysis reports</li>
</ul>
</li>
</ul>
<p>Omix implements these modular steps and displays results in interactive reports.</p>
</div>
<div class="section level2">
<h2 id="benchmark-of-multi-omics-integration-softwares">Benchmark of multi-omics integration softwares<a class="anchor" aria-label="anchor" href="#benchmark-of-multi-omics-integration-softwares"></a>
</h2>
<table class="table">
<colgroup>
<col width="3%">
<col width="11%">
<col width="5%">
<col width="7%">
<col width="15%">
<col width="4%">
<col width="32%">
<col width="9%">
<col width="3%">
<col width="8%">
</colgroup>
<thead><tr class="header">
<th align="left">…1</th>
<th align="left">Supported omics</th>
<th align="left">Pre-processing</th>
<th align="left">Single omic analysis</th>
<th align="left">Integrative models</th>
<th align="left">Use case</th>
<th align="left">Downstream analyses</th>
<th align="left">Interactive visualisations</th>
<th align="left">Language</th>
<th align="left">Ref</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Miodin</td>
<td align="left">SNP, RNA, methylation, Proteins,</td>
<td align="left">+</td>
<td align="left">-</td>
<td align="left">MOFA</td>
<td align="left">BM</td>
<td align="left">-</td>
<td align="left">-</td>
<td align="left">R</td>
<td align="left">Ulfenborg, B. (2019).</td>
</tr>
<tr class="even">
<td align="left">MiBiOmics</td>
<td align="left">miRNA, RNA, Proteins</td>
<td align="left">-</td>
<td align="left">-</td>
<td align="left">Co-inertia analysis</td>
<td align="left">BD</td>
<td align="left">Multi-omics networks</td>
<td align="left">+</td>
<td align="left">Web</td>
<td align="left">(Zoppi et al., 2021)</td>
</tr>
<tr class="odd">
<td align="left">Muon</td>
<td align="left">Single cell omics</td>
<td align="left">-</td>
<td align="left">-</td>
<td align="left">MOFA, WNN</td>
<td align="left">NA</td>
<td align="left">-</td>
<td align="left">-</td>
<td align="left">Python</td>
<td align="left">(Bredikhin et al., 2022)</td>
</tr>
<tr class="even">
<td align="left">Movics</td>
<td align="left">SNP, RNA, methylation, Proteins</td>
<td align="left">-</td>
<td align="left">-</td>
<td align="left">Range of clustering algorithm</td>
<td align="left">STR</td>
<td align="left">Survival analysis, enrichment</td>
<td align="left">-</td>
<td align="left">R</td>
<td align="left">(Lu, 2020)</td>
</tr>
<tr class="odd">
<td align="left">Omix</td>
<td align="left">RNA, Proteins</td>
<td align="left">+</td>
<td align="left">+</td>
<td align="left">MOFA,MEIFESTO, sMBPLS, MBPLS, DIABLO, iCluster</td>
<td align="left">BD, BM, STR</td>
<td align="left">Multi-omics signatures, networks, modules, functional/ cell type/ TF enrichment, pseudotime, etc</td>
<td align="left">+</td>
<td align="left">R</td>
<td align="left">NA</td>
</tr>
</tbody>
</table>
<p>Biomarker Discovery (BD), Biological Mechanisms (BM), Sample stratification (STR)</p>
</div>
<div class="section level2">
<h2 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h2>
<p>The Getting Started section of the documentation contains downloadable examples on how to use Omix.</p>
</div>
<div class="section level2">
<h2 id="data">Data<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<p>The experiments described in our vignettes rely on in house data from the Multi-Omics Atlas Project, which may be obtained from the synapse portal for registered users.</p>
<ul>
<li>Project SynID: syn36812517</li>
<li>Get started data: syn51533729</li>
<li>Pseudo-temporal multi-omics integration: syn51516099</li>
</ul>
</div>
<div class="section level2">
<h2 id="docker-image">Docker image<a class="anchor" aria-label="anchor" href="#docker-image"></a>
</h2>
<p>For reproducibility purposes, we provide a Docker container <a href="https://hub.docker.com/r/eleonoreschneeg/omix" class="external-link">here</a>.</p>
<p>After installing Docker you can first pull the container via:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> pull eleonoreschneeg/omix:latest</span></code></pre></div>
<p>and then run the container:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="ex">docker</span> run <span class="at">--rm</span> <span class="at">-d</span> <span class="at">-v</span> <span class="va">$HOME</span>:/home/rstudio/home <span class="at">-e</span> ROOT=true <span class="at">-e</span> PASSWORD=password <span class="at">-p</span> 8787:8787 eleonoreschneeg/omix:1.0.0</span></code></pre></div>
<p>An RStudio server session can be accessed via a browser at localhost:8787 using Username: rstudio and Password: password Navigate to Omix and open the Omix.Rproj file.</p>
</div>
<div class="section level2">
<h2 id="citation">Citation<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>Please cite <code>Omix</code> as:</p>
<p>Eléonore Schneegans, Nurun Fancy, Michael Thomas, Nanet Willumsen, Paul M Matthews, Johanna Jackson (2023) Omix: A Multi-Omics Integration Pipeline</p>
</div>
<div class="section level2">
<h2 id="contributing">Contributing<a class="anchor" aria-label="anchor" href="#contributing"></a>
</h2>
<p>For feature requests, please open an issue <a href="https://github.com/eleonore-schneeg/Omix/issues" class="external-link">here</a>.</p>
<p>Alternatively, you can fork the repository, add your change and issue a pull request.</p>
</div>

  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small>GPL (&gt;= 3)</small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing Omix</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Eleonore Schneegans <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-6754-6249" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Nurun Fancy <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6481-6266" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.gnu.org/licenses/gpl-3.0" class="external-link"><img src="https://img.shields.io/badge/License-GPLv3-green.svg" alt="License: GPL v3"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Eleonore Schneegans, Nurun Fancy.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
