<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ Omix</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">Omix</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Omix_vignette_CV_MTG.html">Omix - Get started</a>
    </li>
    <li>
      <a href="../articles/Omix_vignette_MAP_CV_pseudotemporal.html">Omix - Pseudotemporal multi-omics integration</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="pre-processing-functions">Pre-processing functions <a href="#pre-processing-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="batch_correction_protein.html">batch_correction_protein()</a></code> </p>
        </td>
        <td><p>Calling different proteomics correction methods.
Internal to <code><a href="../reference/process_protein.html">process_protein()</a></code></p></td>
      </tr><tr><td>
          <p><code><a href="combat_correction.html">combat_correction()</a></code> </p>
        </td>
        <td><p>Proteomics normalisation batch correction using Combat</p></td>
      </tr><tr><td>
          <p><code><a href="filter_protein.html">filter_protein()</a></code> </p>
        </td>
        <td><p>Data filtering function</p></td>
      </tr><tr><td>
          <p><code><a href="generate_multiassay.html">generate_multiassay()</a></code> </p>
        </td>
        <td><p>Generate a MultiAssayExperiment object from multiple single-omics data</p></td>
      </tr><tr><td>
          <p><code><a href="median_centering_correction.html">median_centering_correction()</a></code> </p>
        </td>
        <td><p>Proteomics normalisation batch correction using median centering method</p></td>
      </tr><tr><td>
          <p><code><a href="process_protein.html">process_protein()</a></code> </p>
        </td>
        <td><p>Process bulk proteomics data</p></td>
      </tr><tr><td>
          <p><code><a href="process_rna.html">process_rna()</a></code> </p>
        </td>
        <td><p>Process batch transcriptomics data</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="single-omic-analysis-function">Single-omic analysis function <a href="#single-omic-analysis-function" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="DE_multiomics.html">DE_multiomics()</a></code> </p>
        </td>
        <td><p>Performs DE on multiomics object from the processed slots</p></td>
      </tr><tr><td>
          <p><code><a href="protein_DE_analysis.html">protein_DE_analysis()</a></code> </p>
        </td>
        <td><p>Performs single omic analysis of proteomic layer.</p></td>
      </tr><tr><td>
          <p><code><a href="rna_DE_analysis.html">rna_DE_analysis()</a></code> </p>
        </td>
        <td><p>Performs single omic analysis of transcriptomics layer.</p></td>
      </tr><tr><td>
          <p><code><a href="single_omic_comparisons.html">single_omic_comparisons()</a></code> </p>
        </td>
        <td><p>Compare differential expressed genes and proteins</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="multi-omic-integration-function">Multi-omic integration function <a href="#multi-omic-integration-function" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="getClustNum.html">getClustNum()</a></code> </p>
        </td>
        <td><p>Get estimation of optimal clustering number</p></td>
      </tr><tr><td>
          <p><code><a href="integrate_with_DIABLO.html">integrate_with_DIABLO()</a></code> </p>
        </td>
        <td><p>Vertical integration with DIABLO including model tuning</p></td>
      </tr><tr><td>
          <p><code><a href="integrate_with_MBPLS.html">integrate_with_MBPLS()</a></code> </p>
        </td>
        <td><p>Vertical integration with MBPLS</p></td>
      </tr><tr><td>
          <p><code><a href="integrate_with_MEIFESTO.html">integrate_with_MEIFESTO()</a></code> </p>
        </td>
        <td><p>Vertical integration with MEIFESTO</p></td>
      </tr><tr><td>
          <p><code><a href="integrate_with_MOFA.html">integrate_with_MOFA()</a></code> </p>
        </td>
        <td><p>Vertical integration with MOFA</p></td>
      </tr><tr><td>
          <p><code><a href="integrate_with_iCluster.html">integrate_with_iCluster()</a></code> </p>
        </td>
        <td><p>Vertical integration with iCluster</p></td>
      </tr><tr><td>
          <p><code><a href="integrate_with_sMBPLS.html">integrate_with_sMBPLS()</a></code> </p>
        </td>
        <td><p>Vertical integration with SMBPLS</p></td>
      </tr><tr><td>
          <p><code><a href="vertical_integration.html">vertical_integration()</a></code> </p>
        </td>
        <td><p>Patient specific vertical integration</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="downstream-analysis-of-multi-omics-integrated-data">Downstream analysis of multi-omics integrated data <a href="#downstream-analysis-of-multi-omics-integrated-data" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="clustering_DE_analysis.html">clustering_DE_analysis()</a></code> </p>
        </td>
        <td><p>Performs differential expression between clusters</p></td>
      </tr><tr><td>
          <p><code><a href="integrative_results_supervised.html">integrative_results_supervised()</a></code> </p>
        </td>
        <td><p>Downstream analysis for integrated multi-omics data</p></td>
      </tr><tr><td>
          <p><code><a href="integrative_results_unsupervised.html">integrative_results_unsupervised()</a></code> </p>
        </td>
        <td><p>Downstream analysis for integrated multi-omics data</p></td>
      </tr><tr><td>
          <p><code><a href="multiomics_modules.html">multiomics_modules()</a></code> </p>
        </td>
        <td><p>Compute the module eigen value (PC1 of scaled transcriptomics/proteomics
expression) and correlates it to chosen covariates</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="pseudotime-inference-functions">Pseudotime inference functions <a href="#pseudotime-inference-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="pseudotime_inference.html">pseudotime_inference()</a></code> </p>
        </td>
        <td><p>Pseudotime_inference based on integrated embeddings and clusters</p></td>
      </tr><tr><td>
          <p><code><a href="pseudotime_ordering_modules.html">pseudotime_ordering_modules()</a></code> </p>
        </td>
        <td><p>Generates a plot relating modules eigenvalue along pseudotime</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="helper-functions">Helper functions <a href="#helper-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="ask_chatgpt.html">ask_chatgpt()</a></code> </p>
        </td>
        <td><p>Calls chatGPT with a specific prompt</p></td>
      </tr><tr><td>
          <p><code><a href="counts_rna.html">counts_rna()</a></code> </p>
        </td>
        <td><p>Adds custom processed dataset to multiassay object</p></td>
      </tr><tr><td>
          <p><code><a href="extract_multiomic_signature.html">extract_multiomic_signature()</a></code> </p>
        </td>
        <td><p>Extracts a multi-omics signature from loadings</p></td>
      </tr><tr><td>
          <p><code><a href="extract_weigths.html">extract_weigths()</a></code> </p>
        </td>
        <td><p>Extract_weigths from MOFA model for a given factor of interest at a
defined absolute threshold</p></td>
      </tr><tr><td>
          <p><code><a href="format_res_deseq.html">format_res_deseq()</a></code> </p>
        </td>
        <td><p>Formatting differential expression results from DESEQ2</p></td>
      </tr><tr><td>
          <p><code><a href="format_res_limma.html">format_res_limma()</a></code> </p>
        </td>
        <td><p>Formatting differential expression results from limma</p></td>
      </tr><tr><td>
          <p><code><a href="get_background.html">get_background()</a></code> </p>
        </td>
        <td><p>Get background genes</p></td>
      </tr><tr><td>
          <p><code><a href="get_metadata.html">get_metadata()</a></code> </p>
        </td>
        <td><p>Get metadata</p></td>
      </tr><tr><td>
          <p><code><a href="processed_proteomics.html">processed_proteomics()</a></code> </p>
        </td>
        <td><p>Adds a custom processed proteomics dataset to multiassay object</p></td>
      </tr><tr><td>
          <p><code><a href="processed_rna.html">processed_rna()</a></code> </p>
        </td>
        <td><p>Adds a custom processed transcriptomics slot to multiassay object</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="plotting-functions">Plotting functions <a href="#plotting-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="communities_network.html">communities_network()</a></code> </p>
        </td>
        <td><p>Detects communities in multi-omics network</p></td>
      </tr><tr><td>
          <p><code><a href="community_graph.html">community_graph()</a></code> </p>
        </td>
        <td><p>Creates a subnetwork graph based on detected communities</p></td>
      </tr><tr><td>
          <p><code><a href="interactive_network.html">interactive_network()</a></code> </p>
        </td>
        <td><p>Get interactive network from igraph object</p></td>
      </tr><tr><td>
          <p><code><a href="multiomics_network.html">multiomics_network()</a></code> </p>
        </td>
        <td><p>Get multiomics networks from multiassay object</p></td>
      </tr><tr><td>
          <p><code><a href="multiomics_network_cluster.html">multiomics_network_cluster()</a></code> </p>
        </td>
        <td><p>Creates a multi-omics network based on clustered subgroups following
multi-omics clustering</p></td>
      </tr><tr><td>
          <p><code><a href="multiomics_network_matrix.html">multiomics_network_matrix()</a></code> </p>
        </td>
        <td><p>Get multiomics networks from multimodal object</p></td>
      </tr><tr><td>
          <p><code><a href="plot_OpenTarget.html">plot_OpenTarget()</a></code> </p>
        </td>
        <td><p>Interface to OpenTargets database</p></td>
      </tr><tr><td>
          <p><code><a href="plot_optimal_cluster.html">plot_optimal_cluster()</a></code> </p>
        </td>
        <td><p>Plot optimal number of clusters from <code><a href="../reference/getClustNum.html">getClustNum()</a></code></p></td>
      </tr><tr><td>
          <p><code><a href="volcano_interactive.html">volcano_interactive()</a></code> </p>
        </td>
        <td><p>Builds an interactive volcano plot using plotly</p></td>
      </tr><tr><td>
          <p><code><a href="volcano_interactive_comparison.html">volcano_interactive_comparison()</a></code> </p>
        </td>
        <td><p>Builds an interactive comparison plot using plotly</p></td>
      </tr><tr><td>
          <p><code><a href="volcano_plot_deseq.html">volcano_plot_deseq()</a></code> </p>
        </td>
        <td><p>Builds a volcano plot from DESEQ2 results</p></td>
      </tr><tr><td>
          <p><code><a href="volcano_plot_limma.html">volcano_plot_limma()</a></code> </p>
        </td>
        <td><p>Builds a volcano plot from limma results</p></td>
      </tr></tbody><tbody><tr><th colspan="2">
          <h2 id="functional-enrichment-functions">Functional enrichment functions <a href="#functional-enrichment-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="GSEA.html">GSEA()</a></code> </p>
        </td>
        <td><p>Performs functional gene set enrichment analysis</p></td>
      </tr><tr><td>
          <p><code><a href="cell_type_enrichment.html">cell_type_enrichment()</a></code> </p>
        </td>
        <td><p>Cell type enrichment of communities</p></td>
      </tr><tr><td>
          <p><code><a href="enrichment_custom.html">enrichment_custom()</a></code> </p>
        </td>
        <td><p>Functional enrichment based on given reference annotations</p></td>
      </tr><tr><td>
          <p><code><a href="pathway_analysis_enrichr.html">pathway_analysis_enrichr()</a></code> </p>
        </td>
        <td><p>Functional enrichment analysis using enrichR</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Eleonore Schneegans, Nurun Fancy.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

