#' Downstream analysis for integrated multi-omics data
#'
#' @param multiassay Multiassay experiment object generated by Omix
#' @param integration Possible unsupervised integration methods are `MOFA`,
#' @param dependent Dependent variable for groups.
#' @param correlation_threshold Absolute correlation threshold to draw edges
#' in the network
#' @param disease_id OpenTargets disease id. Default to "MONDO_0004975"
#' (Alzheimer's Disease) Different disease ontologies can be found on
#' https://www.ebi.ac.uk/ols/ontologies
#'
#' @return List object of integrated results
#'
#' @family Multi-omic integration downstream analysis
#'
#' @importFrom purrr possibly
#' @importFrom clusterProfiler enrichGO compareCluster
#' @importFrom enrichplot pairwise_termsim
#' @export

integrative_results_unsupervised <- function(multiassay,
                                             integration = "MOFA",
                                             dependent = "diagnosis",
                                             correlation_threshold = 0.5,
                                             disease_id = "MONDO_0004975") {

  model <- multiassay@metadata$integration[[paste(integration)]]
  covariates <- colnames(samples_metadata(model))

  MOFA2::correlate_factors_with_covariates(model,
                                           covariates = covariates,
                                           plot = "r")
  MOFA2::plot_variance_explained(model, max_r2 = 20)

  selected_factor <- MOFA_get_relevant_factor(MOFAobject = model,
                                              covariate = dependent)

  sign <- MOFA_sign_relevant_factor(model,
    covariate = dependent,
    relevant_factor = selected_factor
  )


  weights <- MOFA2::get_weights(model,
    view = "all",
    factor = "all",
    abs = FALSE,
    scale = FALSE,
    as.data.frame = FALSE
  )

  rna <- data.frame(weights = weights[[1]][, selected_factor])
  protein <- data.frame(weights = weights[[2]][, selected_factor])

  top <- 100
  if (sign == 1) {
    ## high pathology in positive factor (negative corr to outcome)
    rna_positive <- rownames(rna %>% top_n(top))
    rna_negative <- rownames(rna %>% top_n(-top))

    protein_positive <- rownames(protein %>% top_n(top))
    protein_negative <- rownames(protein %>% top_n(-top))
  }

  if (sign == -1) {
    ## high pathology in negative factor (positive corr to outcome)
    rna_positive <- rownames(rna %>% top_n(-top))
    rna_negative <- rownames(rna %>% top_n(top))

    protein_positive <- rownames(protein %>% top_n(-top))
    protein_negative <- rownames(protein %>% top_n(top))
  }

  cli::cli_alert_success("RETRIEVAL OF PROTECTIVE AND DETRIMENTAL
                         MULTI-OMICS SIGNATURE")

  detrimental <- list(
    rna = rna_positive,
    protein = protein_positive
  )

  protective <- list(
    rna = rna_negative,
    protein = protein_negative
  )


  cli::cli_alert_success("BIPARTITE NETWORKS GENERATION")
  protective_network <- multiomics_network(multiassay,
    list = protective,
    correlation_threshold = correlation_threshold
  )


  detrimental_network <- multiomics_network(multiassay,
    list = detrimental,
    correlation_threshold = correlation_threshold
  )


  ### ID MAPPING
  cli::cli_alert_success("ID MAPPING")
  rna_positive <- get_ID_names(rna_positive,
    omic = "rna",
    from = "ensembl_gene_id",
    to = "gene_name"
  )
  rna_negative <- get_ID_names(rna_negative,
    omic = "rna",
    from = "ensembl_gene_id",
    to = "gene_name"
  )
  protein_positive <- get_ID_names(protein_positive,
    omic = "protein",
    from = "uniprot_id",
    to = "gene_name"
  )
  protein_negative <- get_ID_names(protein_negative,
    omic = "protein",
    from = "uniprot_id",
    to = "gene_name"
  )

  detrimental <- list(
    rna = rna_positive,
    protein = protein_positive
  )
  protective <- list(
    rna = rna_negative,
    protein = protein_negative
  )

  cli::cli_alert_success("DETECTION OF NETWORK COMMUNITIES")
  detrimental_communities <- communities_network(detrimental_network)

  protective_communities <- communities_network(protective_network)


  cli::cli_alert_success("CELL TYPE ENRICHMENT OF COMMUNITIES")

  detrimental_cell_type <- cell_type_enrichment(
    multiassay = multiassay,
    communities = detrimental_communities
  )

  protective_cell_type <- cell_type_enrichment(
    multiassay = multiassay,
    communities = protective_communities
  )


  cli::cli_alert_success("FUNCTIONAL ENRICHMENT")
  background <- get_background(multiassay)

  enrichr_detrimental <- lapply(detrimental, pathway_analysis_enrichr)
  enrichr_protective <- lapply(protective, pathway_analysis_enrichr)

  enrichGO_detrimental <- lapply(
    detrimental,
    function(x) {
      clusterProfiler::enrichGO(
        gene = x,
        OrgDb = org.Hs.eg.db::org.Hs.eg.db,
        universe = background,
        keyType = "SYMBOL",
        ont = "BP",
        pAdjustMethod = "BH",
        pvalueCutoff = 0.05,
        qvalueCutoff = 0.05,
        minGSSize = 2,
        maxGSSize = 500
      )
    }
  )

  enrichGO_protective <- lapply(protective, function(x) {
    clusterProfiler::enrichGO(
      gene = x,
      OrgDb = org.Hs.eg.db::org.Hs.eg.db,
      universe = background,
      keyType = "SYMBOL",
      ont = "BP",
      pAdjustMethod = "BH",
      pvalueCutoff = 0.05,
      qvalueCutoff = 0.05,
      minGSSize = 2,
      maxGSSize = 500
    )
  })


  #### Communities comparisons / mix proteomics transcriptomics if highly corr
  detrimental_comp <- clusterProfiler::compareCluster(
    geneCluster =
      detrimental_communities,
    fun = "enrichGO",
    OrgDb = "org.Hs.eg.db",
    keyType = "SYMBOL",
    universe = background,
    ont = "BP",
    minGSSize = 2,
    pAdjustMethod = "BH",
    pvalueCutoff = 0.05,
    qvalueCutoff = 0.05
  )


  detrimental_comp <- enrichplot::pairwise_termsim(detrimental_comp)

  protective_comp <- clusterProfiler::compareCluster(
    geneCluster =
      protective_communities,
    fun = "enrichGO",
    OrgDb = "org.Hs.eg.db",
    keyType = "SYMBOL",
    universe = background,
    ont = "BP",
    minGSSize = 2,
    pAdjustMethod = "BH",
    pvalueCutoff = 0.05,
    qvalueCutoff = 1
  )

  protective_comp <- enrichplot::pairwise_termsim(protective_comp)


  ### OPEN TARGETS
  cli::cli_alert_success("TESTING MULTI_OMICS SIGNATURE ASSOCIATION
                           WITH DISEASE OF INTEREST VIA THE OPENTARGETS API")

  opentarget_detrimental <- plot_OpenTarget(
    disease_id = disease_id,
    size = 1000,
    genes = unlist(detrimental)
  )
  opentarget_protective <- plot_OpenTarget(
    disease_id = disease_id,
    size = 1000,
    genes = unlist(protective)
  )



  #### RESULTS OBJECT
  integrative_results <- list(
    detrimental = list(
      multiomics_signature = detrimental,
      network = detrimental_network,
      network_communities = detrimental_communities,
      cell_type_communities = detrimental_cell_type,
      enrichment = enrichr_detrimental,
      open_targets = opentarget_detrimental
    ),
    protective = list(
      multiomics_signature = protective,
      network = protective_network,
      network_communities = protective_communities,
      cell_type_communities = protective_cell_type,
      enrichment = enrichr_protective,
      open_targets = opentarget_protective
    ),
    comparison = list(
      detrimental = detrimental_comp,
      protective = protective_comp
    )
  )


  return(integrative_results)
}


#' Selects the top factor that is most strongly associated with outcomes of interest

#'
#' @param MOFAobject MOFA model
#' @param covariate Covariate of interest
#'
#' @return String of relevant factor
#' @importFrom MOFA2 correlate_factors_with_covariates
#' @export

MOFA_get_relevant_factor <- function(MOFAobject, covariate) {
  corr <- MOFA2::correlate_factors_with_covariates(MOFAobject,
    covariates = covariate,
    plot = "r",
    return_data = TRUE
  )


  relevant_factor <- rownames(corr)[which(abs(corr) == max(abs(corr)))]
  relevant_factor_number <- as.numeric(substr(relevant_factor, 7, 9))

  return(relevant_factor_number)
}

MOFA_sign_relevant_factor <- function(MOFAobject, covariate, relevant_factor) {
  corr <- MOFA2::correlate_factors_with_covariates(MOFAobject,
    covariates = covariate,
    plot = "r",
    return_data = TRUE
  )
  sign <- sign(corr[relevant_factor, ])
  return(sign)
}
