#' Generate multi-omics QC results
#'
#' @param multiassay Multiassay experiment object generated by Omix
#' `generate_multiassay`function
#' 
#' @importFrom MultiAssayExperiment intersectRows assays wideFormat
#' @importFrom stats cor
#'
#' @return parameters$processing_outputs$multiomics slot
#' @family Report
#' @export
#'
multi_omics_qc <- function(multiassay) {

  processing_outputs=list()
  suppressMessages({
  # Transcriptomics & proteomics slots
  rownames(multiassay@ExperimentList@listData[["rna_processed"]]) <-make.unique(get_ID_names(multiassay,
                                                              id=rownames(multiassay@ExperimentList@listData[["rna_processed"]]),
                                                              omic = "rna",
                                                              from = "ensembl_gene_id",
                                                              to = "gene_name"))
  multi <- multiassay[, , c("rna_processed", "protein_processed")] #
  CompleteMulti <- multi[, complete.cases(multi), ] #
  CommonGenes_Samples <-   MultiAssayExperiment::intersectRows(CompleteMulti) #
  metadata_intersection <- data.frame(colData(CompleteMulti)) #


  dim_rna_raw=dim(multiassay@ExperimentList@listData[["rna_raw"]])
  dim_rna_processed=dim(multiassay@ExperimentList@listData[["rna_processed"]])
  dim_protein_raw=dim(multiassay@ExperimentList@listData[["protein_raw"]])
  dim_protein_processed=dim(multiassay@ExperimentList@listData[["protein_processed"]])

  dimensions=list(dim_rna_raw=dim_rna_raw,
       dim_rna_processed=dim_rna_processed,
       dim_protein_raw=dim_protein_raw,
       dim_protein_processed=dim_protein_processed)
    # Intersection of samples & genes
  df <- data.frame(
    mRNA = rbind(
      dim(CompleteMulti@ExperimentList@listData[[1]])[1],
      dim(CommonGenes_Samples@ExperimentList@listData[[1]])[1]
    ),
    proteins = rbind(
      dim(CompleteMulti@ExperimentList@listData[[2]])[1],
      dim(CommonGenes_Samples@ExperimentList@listData[[2]])[1]
    )
  )

  df$type <- c("before intersection", "intersection")
  long <- melt(df) #
  colnames(long) <- c("state", "omic", "# genes")


  ## Correlation of mRNA abundance and protein expression

  subacc <- CommonGenes_Samples
  subacc.list <- MultiAssayExperiment::assays(subacc)
  subacc.list <- lapply(subacc.list, t)
  corres <- stats::cor(subacc.list[[1]], subacc.list[[2]])
  gene_max_cor <- names(which.max(diag(corres)))
  df <- MultiAssayExperiment::wideFormat(subacc[gene_max_cor, , ])
  df2 <- data.frame(df@listData) #

  gene_min_cor <- names(which.min(diag(corres)))
  df <- MultiAssayExperiment::wideFormat(subacc[gene_min_cor, , ])
  df3 <- data.frame(df@listData) #

  correlations <- data.frame(corr = diag(corres))
  correlations$gene <- rownames(correlations)
  correlations <- correlations[order(correlations$corr), ]
  correlations$gene <- factor(correlations$gene, levels = correlations$gene)


  mRNA <- df2[, 2]
  protein <- df2[, 3]
  max_cor <- stats::cor(mRNA, protein, method = c("pearson")) #

  mRNA <- df3[, 2]
  protein <- df3[, 3]
  min_cor <- stats::cor(mRNA, protein, method = c("pearson")) #


  getLoadings <- function(x, ncomp = 10, dolog = FALSE, center = TRUE, scale. = TRUE) {
    if (dolog) {
      x <- log2(x + 1)
    }
    pc <- stats::prcomp(x, center = center, scale. = scale.)
    return(t(pc$rotation[, 1:10]))
  }

  subacc <- c(subacc, rnaseqPCA = getLoadings(MultiAssayExperiment::assays(subacc)[[1]], dolog = TRUE), mapFrom = 1L)
  subacc <- c(subacc, proteinPCA = getLoadings(MultiAssayExperiment::assays(subacc)[[2]], center = FALSE, scale. = FALSE), mapFrom = 2L)

  pca <- subacc[, , 3:4]
  df_pca <- MultiAssayExperiment::wideFormat(pca)[, -1]
  mycors <- stats::cor(as.matrix(df_pca))
  mycors <- abs(mycors)
  diag(mycors) <- NA
  has.high.cor <- apply(mycors, 2, max, na.rm = TRUE) > 0
  mycors <- mycors[has.high.cor, has.high.cor] #

  summary=list(capture.output(experiments(multiassay)))
  plot=  MultiAssayExperiment::upsetSamples(multiassay)
  processing_outputs=list(dimensions= dimensions,
                          intersection_common_features=long, #Intersection of samples & genes
                          correlations= correlations,
                          gene_max_cor=gene_max_cor,
                          gene_min_cor=  gene_min_cor,
                          feature_max_cor=df2,
                          feature_min_cor=df3,
                          max_cor=max_cor,
                          min_cor=  min_cor,
                          pca_cor=mycors, # rna / prot PCA
                          plot=plot,
                          summary=summary)

  MultiAssayExperiment::metadata(multiassay)$parameters$processing_outputs$multiomics <- processing_outputs
  })
  return(multiassay)
}

