################################################################################
#' Performs single omic analysis of transcriptomics layer.
#'
#' @param multiassay Multiassay object generated by Omix
#' @param dependent Dependent variable in the differential expression analysis
#' @param levels Levels for dependent variable. Control group should be put
#' first.
#' @param filter_protein_coding Logical whether to filter out non protein coding
#' genes
#' @param log2FoldChange log2 fold-change cutoff.
#' @param padj adjusted p value threshold. Default to 0.05
#' @return Adds differential expression results to Multiassay object generated
#' by Omix
#'
#' @family Single-omic
#'
#' @importFrom dplyr filter arrange
#' @importFrom DESeq2 DESeq results
#' @importFrom MultiAssayExperiment metadata
#' @export
#'
rna_DE_analysis <- function(multiassay,
                            dependent = "diagnosis",
                            levels = NULL,
                            filter_protein_coding = TRUE,
                            log2FoldChange = 0.5,
                            padj=0.05) {
  if (("dds" %in% names(multiassay@metadata)) == FALSE) {
    stop(cli::cli_alert_danger(
      paste("dds object not found in metadata, please run",
        cli::style_bold("process_rna"),
        sep = " "
      )
    ))
  }

  dds <- multiassay@metadata[["dds"]]
  dds <- DESeq2::DESeq(dds)
  res <- list()

  if (!is.numeric(dds[[dependent]])) {
    if (!is.null(levels)) {
      dds[[dependent]] <- factor(dds[[dependent]], levels = levels)
    }
    perCombs <- split(
      t(combn(levels(dds[[dependent]]), 2)),
      seq(nrow(t(combn(levels(dds[[dependent]]), 2))))
    )

    names <- list()
    for (i in 1:length(perCombs)) {
      names[[i]] <- c(perCombs[[i]][2], perCombs[[i]][1])
    }

    list_contrast <- lapply(names, function(x) {
      c(dependent, x)
    })
    res <- lapply(list_contrast, function(x) {
      DESeq2::results(dds,
        lfcThreshold = 0,
        alpha = 0.05,
        contrast = x,
        independentFiltering = TRUE
      )
    })
  }

  names2 <- list()
  for (i in 1:length(perCombs)) {
    names2[i] <- paste0(perCombs[[i]][2], "vs", perCombs[[i]][1])
  }

  names(res) <- unlist(names2)

  gene_id_conversion <- as.data.frame(
    multiassay@ExperimentList@listData[["rna_raw"]]@elementMetadata@listData)

  res <- lapply(res, format_res_deseq,
    n_label = 20,
    gene_id_conversion = gene_id_conversion,
    filter_protein_coding = filter_protein_coding,
    log2FoldChange = log2FoldChange
  )

  list_DEG_limma <- list()
  list_DEG_limma <- lapply(res, function(x) {
    up <- x$gene_name[which(x$padj <= padj & x$de == "Up")]
    down <- x$gene_name[which(x$padj <= padj & x$de == "Down")]
    return(list(
      up = up,
      down = down
    ))
  })
  names(list_DEG_limma) <- unlist(names2)

  res$plot <- lapply(res, function(x) {
    volcano_plot_deseq(
      dt = x,
      log2FoldChange = log2FoldChange
    )
  })

  res$sig_gene <- list_DEG_limma

  functional_list=list()
  for(j in names(res$sig_gene)){
    functional_list[[j]]=lapply(res$sig_gene[[j]] ,pathway_analysis_enrichr)
  }
  res$functional_enrichment <- functional_list
  MultiAssayExperiment::metadata(multiassay)$DEG <- res
  MultiAssayExperiment::metadata(multiassay)$parameters$single_omic$de$rna <- list(dependent = dependent,
                                                                                design= multiassay@metadata[["dds"]]@design,
                                                                                levels = levels,
                                                                                log2FoldChange = log2FoldChange,
                                                                                padj=padj,
                                                                                filter_protein_coding =filter_protein_coding)


  return(multiassay)
}
