#' Performs WGCNA on processed slots (rna and protein)
#'
#' @param multiassay Multiassay object generated by Omix
#' @param dependent Continuous dependent variable for sparse partial least squares
#' @param covariates clinical annotations to be correlated with modules
#' @return WGCNA for processed transcriptomics and proteomics
#' @import WGCNA
#' 
#'
#' @family Single-omic
#'
wgcna_wrapper <- function(multiassay,
                          dependent, 
                          covariates){
  

  args= list(multiassay=multiassay,
             slots = c(
               "rna_processed",
               "protein_processed"),
             intersect_genes = FALSE,
             ID_type = "gene_name")
  suppressMessages({
    multimodal <- do.call(get_multimodal_object, args)
  })
  multimodal_omics <- multimodal[[1]]
  
  metadata <- multimodal[[2]]

  net = blockwiseModules(t(multimodal_omics$mRNA), power = 8,
                         deepSplit=4,
                         TOMType = "signed", minModuleSize = 30,
                         reassignThreshold = 0, mergeCutHeight = 0.1,
                         numericLabels = TRUE, pamRespectsDendro = FALSE,
                         saveTOMs = FALSE,
                         verbose = 3,
                         networkType="signed hybrid")
  
  moduleColors = labels2colors(net$colors)
  moduleLabels = net$colors

  
  nGenes = nrow(multimodal_omics$mRNA);
  nSamples = ncol(multimodal_omics$mRNA);
  # Recalculate MEs with color labels
  MEs0 = moduleEigengenes(t(multimodal_omics$mRNA), moduleColors)$eigengenes
  MEs = orderMEs(MEs0)
  moduleTraitCor = cor(MEs,   metadata[,covariates], use = "p");
  moduleTraitPvalue = corPvalueStudent(moduleTraitCor, nSamples)
  
  corrplot=corrplot::corrplot(t(moduleTraitCor[1:nrow(moduleTraitCor - 1), ]),
                              p.mat = t(moduleTraitPvalue),
                              insig = "label_sig",
                              sig.level = c(0.001, 0.01, 0.05),
                              diag = TRUE,
                              tl.cex=0.5,
                              pch.cex=0.5,
                              col = colorRampPalette(c("#2166AC","#4393C3","#92C5DE","#D1E5F0","#FDDBC7","#F4A582", "#D6604D","#B2182B"))(50))
  
  
  
  # Define variable weight containing the weight column of datTrait
  traits = metadata;
  names(traits) = colnames(metadata)
  # names (colors) of the modules
  modNames = substring(names(MEs), 3)
  geneModuleMembership = as.data.frame(cor(t(multimodal_omics$mRNA), MEs, use = "p"));
  
  MMPvalue = as.data.frame(corPvalueStudent(as.matrix(geneModuleMembership), nSamples));
  names(geneModuleMembership) = paste("MM", modNames, sep="");
  names(MMPvalue) = paste("p.MM", modNames, sep="");
  
  geneTraitSignificance = as.data.frame(cor(t(multimodal_omics$mRNA),traits, use = "p"));
  GSPvalue = as.data.frame(corPvalueStudent(as.matrix(geneTraitSignificance), nSamples));
  
  names(geneTraitSignificance) = paste("GS.", names(traits), sep="");
  names(GSPvalue) = paste("p.GS.", names(traits), sep="")
  
  
  ###
  module = "black"
  trait_interest="amyloid"
  column = match(module, modNames);
  moduleGenes = moduleColors==module
  geneModuleMembership[moduleGenes, column]
  geneTraitSignificance[moduleGenes, paste0("GS.",trait_interest)]
  
  
  df=data.frame('Module_Membership'=geneModuleMembership[moduleGenes, column],'Gene_significance'=geneTraitSignificance[moduleGenes, paste0("GS.",trait_interest)],"Gene"=rownames(geneModuleMembership[moduleGenes,]))
  
  
  return(list(
    metadata_modules = metadata,
    corrplot = corrplot,
    moduleTraitCor = moduleTraitCor,
    moduleTraitPvalue = moduleTraitPvalue,
    moduleColors = moduleColors,
    modulesLabels = df$feature,
    modules_eigen_value = MEs0
  ))
  
}

